[PATCH v5] IPsec validation test application fixes

asasidharan opened the pull request:
This PR consists of following two fixes to IPsec validation test applications:

1.  Fix test_out_ipv4_esp_null_aes_gmac_128() test function:
Perform both outbound and inbound processing similar to test_out_ipv4_esp_aes_gcm128. This is required as AES-GMAC algorithm uses IV for ICV generation and is generated by the ODP IPsec implementation during the IPsec outbound processing. The IV value is not known to the receiver until it receives the packet, but the test function expects a predefined IV value in the resulting packet.

2. Set L3 type for a test packet in ipsec_packet()
Similar to setting the l3 offset, set the L3 type to IPv4/IPv6 in ipsec_packet()

JannePeltonen commented on test/validation/api/ipsec/ipsec.c:
```diff
@@ -447,6 +449,12 @@ odp_packet_t ipsec_packet(const ipsec_test_packet *itp)
+	eth_hdr = (odph_ethhdr_t *)odp_packet_l2_ptr(pkt, NULL);
+	type = odp_be_to_cpu_16(eth_hdr->type);
+	if (type == ODPH_ETHTYPE_IPV4)
+		odp_packet_has_ipv4_set(pkt, 1);
```
This is not right. IPsec operations do not require the has_ipv4 or has_ipv6 flags to be set (see the api spec, e.g. ipsec.h:1287). The implementation is expected to look into the packet data to see whether the packet is IPv4 or IPv4. Setting the has_ipv4 or has_ipv6 flags here would make the validation test accept incorrect implementations that require the flags to be set according to the real packet type.

asasidharan replied:
Abandoning this change as pointed out, as per the spec only l3 and l4 offsets need to have valid values.

JannePeltonen commented:
The AES-GMAC fix looks good.

The problem is actually reported in issue #800. There is similar problem in  test_out_ipv4_esp_null_aes_gmac_128(). Could you perhaps fix that too? And please add this tag in the commit log: Fixes: https://github.com/OpenDataPlane/odp/issues/800

Please remove the second commit (the has_ipv4 & has_ipv6 addition) from this PR.

asasidharan commented:
> The problem is actually reported in issue #800. There is similar problem in test_out_ipv4_esp_null_aes_gmac_128(). Could you perhaps fix that too? And please add this tag in the commit log: Fixes: #800
> 
> Please remove the second commit (the has_ipv4 & has_ipv6 addition) from this PR.

Updated the pull request with fix for AES-GMAC test functions in both ESP and AH mode along with the tag "Fixes: #800". Also removed the second commit as suggested.

matiaselo commented:
Hi @asasidharan,

Please add Janne's reviewed-by tag to the commit message and do a rebase. After this I can merge the PR.
